paramiko/kex_curve25519.py

```diff
@@ -0,0 +1,218 @@
+# Copyright (C) 2003-2007  Robey Pointer <robeypointer@gmail.com>
+#
+# This file is part of paramiko.
+#
+# Paramiko is free software; you can redistribute it and/or modify it under the
+# terms of the GNU Lesser General Public License as published by the Free
+# Software Foundation; either version 2.1 of the License, or (at your option)
+# any later version.
+#
+# Paramiko is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+# details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Paramiko; if not, write to the Free Software Foundation, Inc.,
+# 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
+
+"""
+Key exchange using DJB's Curve25519. Originally introduced in OpenSSH 6.5, and
+the only kex currently (August 2018) recommended by arthepsy's ssh-audit.
+"""
+
+# Author: Dan Fuhry <dan@fuhry.com>
+
+from hashlib import sha256
+
+from paramiko.message import Message
+from paramiko.py3compat import byte_chr, long
+from paramiko.ssh_exception import SSHException
+from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat
+from cryptography.exceptions import UnsupportedAlgorithm
+from binascii import hexlify
+
+# x25519 was added in cryptography 2.5, but we support as low as 1.5. Trap any
```

Comment:
also minor: x25519 was added in cryptography 2.0 (and this code also depends on features added in cryptography 2.5, but this bit of the comment might fit better in `is_supported()`)